### ISSUE TITLE
`terra::rast(type = "xyz")` doesn't respect derived resolution when extent is larger than what is represented by input data.frame

### ISSUE THREAD

**Issue by mikoontz-vp** [rspatial/terra#1070]
If you have a sparse representation of a raster in data.frame format and you want to convert it back to a `SpatRaster`, the resolution can be correctly derived, but is ultimately changed in the end to accommodate expanding the extent.

MWE (riffing on the example in the ?terra::rast() help file):
```
# Original example in the help file
# from a matrix 
m <- matrix(1:25, nrow=5, ncol=5)
rm <- rast(m)

# from a "xyz" data.frame
d <- as.data.frame(rm, xy=TRUE)
head(d)
rast(d, type="xyz")

# Additions to the example:
# sparse data frame 
d_sparse <- d[1:10, ]
head(d_sparse)
# If I want to keep the overall extent, the correctly derived resolution of c(1, 1) changes to c(1, 2.5)
rast(d_sparse, type="xyz", extent = ext(rm))
```

This happens because the check on whether or not an extent has been passed to `rast()` (https://github.com/rspatial/terra/blob/867f9dd202b9559f5128a98ceea15c5287a15fb6/R/rast.R#L409) comes after some of the direct assignments of attributes of the template raster (`r`) like `res()` and `ext()`.

This appears to work better if the check for a user-provided extent comes earlier, so the end of the method might look like:
```
d <- dim(xyz)

if (!is.null(extent)) {
  r <- rast(extent = extent, resolution = c(rx, ry), crs = crs, nlyrs = d[2]-2)
} else {
  minx <- min(x) - 0.5 * rx
  maxx <- max(x) + 0.5 * rx
  miny <- min(y) - 0.5 * ry
  maxy <- max(y) + 0.5 * ry
  
  r <- rast(xmin=minx, xmax=maxx, ymin=miny, ymax=maxy, crs=crs, nlyrs=d[2]-2)
  res(r) <- c(rx, ry)
  ext(r) <- round(ext(r), digits+2)
}

cells <- cellFromXY(r, xyz[,1:2])
if (d[2] > 2) {
  names(r) <- ln[-c(1:2)]
  v <- try( matrix(NA, nrow=ncell(r), ncol= nlyr(r)) )
  if (inherits(v, "try-error")) {
    error(paste("cannot make matrix with ", ncell(r), " rows"))
  }
  v[cells, ] <- xyz[, -c(1:2)]
  values(r) <- v
}

return(r)
```

This didn't quite fix my use case, because the `v <- try( matrix(NA, nrow=ncell(r), ncol= nlyr(r)) )` line wasn't memory safe and won't proceed in chunks as written, but that's for another issue if I can't figure it out!

**Comment by Rapsodia86** [comment]
See here: https://rdrr.io/cran/terra/man/inplace.html

**Comment by mikoontz** [comment]
Yes, thanks! I read the help file after Robert mentioned the function. It might be worth some extra documentation in the "See also" section of `values()`, `values()<-`, and `setValues()` to help users navigate their various uses. I don't see anything in the help file for `set.values()` that mentions memory issues with large rasters though, so this still might be worth another issue.

**Comment by kadyb** [comment]
> I don't see anything in the help file for set.values() that mentions memory issues with large rasters though, so this still might be worth another issue.

Do you have enough memory to hold this raster in memory? This code works ok for me on 64 GB RAM:

```r
terraOptions(memfrac = 0.9)
n = 30000
r = rast(nrows = n, ncols = n)
v = rep.int(0, times = n^2)
c = seq_len(ncell(r))
set.values(r, cells = c, values = v)
```

Maybe you should initialize the raster on disk (not in memory) using [init()](https://rspatial.github.io/terra/reference/init.html)?

**Comment by mikoontz** [comment]
It fails even if using `init()`. The raster is ~16 billion cells (so a couple orders of magnitude larger than the example in your code snippet). The following fails:

```
library(terra)

r <- rast(nrow = 100000, ncol = 100000)
r <- init(x = r, 0)

set.seed(1)
idx <- sample(x = ncell(r), size = 1000)
set.values(x = r, cells = idx, values = 1)
```

This comment (https://github.com/rspatial/terra/issues/532#issuecomment-1089748463) suggests that perhaps `set.values()` isn't well-supported for very large rasters, which may be an intentional design choice.

**Comment by kadyb** [comment]
Pretty large raster. It seems like over 70 GB is needed to process it in memory.

```r
mem_info(r)
#> needed (n=1)    : 74.51
```

I tried using `r[idx] <- 1` but after 15 min it still didn't finish.

**Comment by mikoontz** [comment]
Okay I learned that I might be able to update specific cells in a very large raster if it is on disk using raster::update(). It doesn't look like updating cell values is supported yet in the terra::update() version. 

https://rdrr.io/cran/raster/man/update.html

**Comment by mikoontz-vp** [comment]
Whoops, never mind! `raster::update()` for specific cell values is no longer supported (looking at the code, it seems it relied heavily on {rgdal} so not surprising that support has been removed since {rdgal} has been retired.) I think we can safely close this issue now and I'll focus on other possible solutions suggested by folks.

@rhijmans: if you were referring to the original code snippet of mine in this thread as not being generally optimal, then that might warrant further discussion here. The creation of the new matrix of cells, which then get populated in the correct index locations by the `xyz` data.frame, came from the `rast()` codebase. Are you suggesting that `rast()` itself could be simplified in the `type=xyz` case by using `set.values()`? Something like:

```
cells <- cellFromXY(r, xyz[,1:2])
if (d[2] > 2) {
  names(r) <- ln[-c(1:2)]
  
  # Pseudocode below to account for setting values on multiple layers
  # Loop over the 3rd through final column of xyz , setting each layer of r to the jth column values
  for (j in 3:d[2]) {
    # memory check here?  
    set.values(r, cells = cells, values = xyz[, j], layer = j)
  }
  
  return(r)
```

**Comment by rhijmans** [comment]
@mikoontz: 
- please do raise a new issue. I can make `x[i] <- v`  and `set.values` memory safe.
- `update` was removed from raster as it depends on terra for those things, and terra does not have that yet.
- I do not think it make sense to let `rast(xyz=TRUE)` figure out the geometry if you already have the geometry; and thus can simply use cell numbers instead of coordinates. It is meant for a situation where you get a file with xyz and you want to derive  the raster geometry from that.
